README.md
# extended_image

[![pub package](https://img.shields.io/pub/v/extended_image.svg)](https://pub.dartlang.org/packages/extended_image) [![GitHub stars](https://img.shields.io/github/stars/fluttercandies/extended_image)](https://github.com/fluttercandies/extended_image/stargazers) [![GitHub forks](https://img.shields.io/github/forks/fluttercandies/extended_image)](https://github.com/fluttercandies/extended_image/network) [![GitHub license](https://img.shields.io/github/license/fluttercandies/extended_image)](https://github.com/fluttercandies/extended_image/blob/master/LICENSE) [![GitHub issues](https://img.shields.io/github/issues/fluttercandies/extended_image)](https://github.com/fluttercandies/extended_image/issues) <a target="_blank" href="https://jq.qq.com/?_wv=1027&k=5bcc0gy"><img border="0" src="https://pub.idqqimg.com/wpa/images/group.png" alt="flutter-candies" title="flutter-candies"></a>

Language: [English](README.md) | [中文简体](README-ZH.md)

A powerful official extension library of image, which support placeholder(loading)/ failed state, cache network, zoom pan image, photo view, slide out page, editor(crop,rotate,flip), paint custom etc.

[Web demo for ExtendedImage](https://fluttercandies.github.io/extended_image/)

## Table of contents

- [extended_image](#extended_image)
  - [Table of contents](#table-of-contents)
  - [Cache Network](#cache-network)
    - [Simple use](#simple-use)
    - [Use ExtendedNetworkImageProvider](#use-extendednetworkimageprovider)
  - [Load State](#load-state)
    - [demo code](#demo-code)
  - [Zoom Pan](#zoom-pan)
    - [double tap animation](#double-tap-animation)
  - [Editor](#editor)
    - [crop aspect ratio](#crop-aspect-ratio)
    - [crop,flip,reset](#cropflipreset)
    - [crop data](#crop-data)
      - [dart library(stable)](#dart-librarystable)
      - [native library(faster)](#native-libraryfaster)
  - [Photo View](#photo-view)
  - [Slide Out Page](#slide-out-page)
    - [enable slide out page](#enable-slide-out-page)
    - [include your page in ExtendedImageSlidePage](#include-your-page-in-extendedimageslidepage)
    - [make sure your page background is transparent](#make-sure-your-page-background-is-transparent)
    - [push with transparent page route](#push-with-transparent-page-route)
  - [Border BorderRadius Shape](#border-borderradius-shape)
  - [Clear Save](#clear-save)
    - [clear](#clear)
    - [save network](#save-network)
  - [Show Crop Image](#show-crop-image)
  - [Paint](#paint)
  - [WaterfallFlow](#waterfallflow)
  - [CollectGarbage/viewportBuilder](#collectgarbageviewportbuilder)
  - [Other APIs](#other-apis)

## Cache Network

### Simple use

You can use `ExtendedImage.network` as Image Widget

```dart
ExtendedImage.network(
  url,
  width: ScreenUtil.instance.setWidth(400),
  height: ScreenUtil.instance.setWidth(400),
  fit: BoxFit.fill,
  cache: true,
  border: Border.all(color: Colors.red, width: 1.0),
  shape: boxShape,
  borderRadius: BorderRadius.all(Radius.circular(30.0)),
  //cancelToken: cancellationToken,
)
```

### Use ExtendedNetworkImageProvider

[ExtendedNetworkImageProvider](https://github.com/fluttercandies/extended_image_library/blob/master/lib/src/extended_network_image_provider.dart)

```dart
   ExtendedNetworkImageProvider(
      this.url, {
      this.scale = 1.0,
      this.headers,
      this.cache: false,
      this.retries = 3,
      this.timeLimit,
      this.timeRetry = const Duration(milliseconds: 100),
      CancellationToken cancelToken,
    })  : assert(url != null),
          assert(scale != null),
          cancelToken = cancelToken ?? CancellationToken();
```

| parameter   | description                                                                           | default             |
| ----------- | ------------------------------------------------------------------------------------- | ------------------- |
| url         | The URL from which the image will be fetched.                                         | required            |
| scale       | The scale to place in the [ImageInfo] object of the image.                            | 1.0                 |
| headers     | The HTTP headers that will be used with [HttpClient.get] to fetch image from network. | -                   |
| cache       | whether cache image to local                                                          | false               |
| retries     | the time to retry to request                                                          | 3                   |
| timeLimit   | time limit to request image                                                           | -                   |
| timeRetry   | the time duration to retry to request                                                 | milliseconds: 100   |
| cancelToken | token to cancel network request                                                       | CancellationToken() |

you can create new provider and extends it with ExtendedProvider, and override instantiateImageCodec method.
so that you can handle image raw data here (compress image).

## Load State

Extended Image provide 3 states(loading,completed,failed), you can define your state widget with
loadStateChanged call back.

loadStateChanged is not only for network, if your image need long time to load,
you can set enableLoadState(default value is true for network and others are false) to true

![img](https://github.com/fluttercandies/Flutter_Candies/blob/master/gif/extended_image/custom.gif)

Notice:

- if you don't want to override any state, please return null in this case

- if you want to override size or sourceRect, you can override it with ExtendedRawImage at completed state

- if you want to add something (like animation) at completed state, you can override it with ExtendedImageState.completedWidget

- ExtendedImageState.completedWidget is include gesture or editor, so that you would't miss them

```dart
/// custom load state widget if you want
    final LoadStateChanged loadStateChanged;

enum LoadState {
  //loading
  loading,
  //completed
  completed,
  //failed
  failed
}

  ///whether has loading or failed state
  ///default is false
  ///but network image is true
  ///better to set it's true when your image is big and take some time to ready
  final bool enableLoadState;
```

ExtendedImageState(LoadStateChanged call back)

| parameter/method             | description                                                                                                                                   | default |
| ---------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------- | ------- |
| extendedImageInfo            | image info                                                                                                                                    | -       |
| extendedImageLoadState       | LoadState(loading,completed,failed)                                                                                                           | -       |
| returnLoadStateChangedWidget | if this is true, return widget which from LoadStateChanged function immediately(width/height/gesture/border/shape etc, will not effect on it) | -       |
| imageProvider                | ImageProvider                                                                                                                                 | -       |
| invertColors                 | invertColors                                                                                                                                  | -       |
| imageStreamKey               | key of image                                                                                                                                  | -       |
| reLoadImage()                | if image load failed,you can reload image by call it                                                                                          | -       |
| completedWidget              | return completed widget include gesture or editor                                                                                             | -       |
| loadingProgress              | return the loading progress for network image (ImageChunkEvent )                                                                              | -       |

```dart
abstract class ExtendedImageState {
  void reLoadImage();
  ImageInfo get extendedImageInfo;
  LoadState get extendedImageLoadState;

  ///return widget which from LoadStateChanged function immediately
  bool returnLoadStateChangedWidget;

  ImageProvider get imageProvider;

  bool get invertColors;

  Object get imageStreamKey;

  Widget get completedWidget;
}
```

### demo code

```dart
ExtendedImage.network(
  url,
  width: ScreenUtil.instance.setWidth(600),
  height: ScreenUtil.instance.setWidth(400),
  fit: BoxFit.fill,
  cache: true,
  loadStateChanged: (ExtendedImageState state) {
    switch (state.extendedImageLoadState) {
      case LoadState.loading:
        _controller.reset();
        return Image.asset(
          "assets/loading.gif",
          fit: BoxFit.fill,
        );
        break;

      ///if you don't want override completed widget
      ///please return null or state.completedWidget
      //return null;
      //return state.completedWidget;
      case LoadState.completed:
        _controller.forward();
        return FadeTransition(
          opacity: _controller,
          child: ExtendedRawImage(
            image: state.extendedImageInfo?.image,
            width: ScreenUtil.instance.setWidth(600),
            height: ScreenUtil.instance.setWidth(400),
          ),
        );
        break;
      case LoadState.failed:
        _controller.reset();
        return GestureDetector(
          child: Stack(
            fit: StackFit.expand,
            children: <Widget>[
              Image.asset(
                "assets/failed.jpg",
                fit: BoxFit.fill,
              ),
              Positioned(
                bottom: 0.0,
                left: 0.0,
                right: 0.0,
                child: Text(
                  "load image failed, click to reload",
                  textAlign: TextAlign.center,
                ),
              )
            ],
          ),
          onTap: () {
            state.reLoadImage();
          },
        );
        break;
    }
  },
)
```

## Zoom Pan

![img](https://github.com/fluttercandies/Flutter_Candies/blob/master/gif/extended_image/zoom.gif)

ExtendedImage

| parameter                | description                                                                     | default |
| ------------------------ | ------------------------------------------------------------------------------- | ------- |
| mode                     | image mode (none, gesture, editor)                                              | none    |
| initGestureConfigHandler | init GestureConfig when image is ready，for example, base on image width/height | -       |
| onDoubleTap              | call back of double tap under ExtendedImageMode.gesture                         | -       |
| extendedImageGestureKey  | you can handle zoom/pan by using this key manually                              | -       |

GestureConfig

| parameter         | description                                                                                                                                                          | default                 |
| ----------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------- |
| minScale          | min scale                                                                                                                                                            | 0.8                     |
| animationMinScale | the min scale for zooming then animation back to minScale when scale end                                                                                             | minScale \_ 0.8         |
| maxScale          | max scale                                                                                                                                                            | 5.0                     |
| animationMaxScale | the max scale for zooming then animation back to maxScale when scale end                                                                                             | maxScale \_ 1.2         |
| speed             | speed for zoom/pan                                                                                                                                                   | 1.0                     |
| inertialSpeed     | inertial speed for zoom/pan                                                                                                                                          | 100                     |
| cacheGesture      | save Gesture state (for example in ExtendedImageGesturePageView, gesture state will not change when scroll back),**remember clearGestureDetailsCache at right time** | false                   |
| inPageView        | whether in ExtendedImageGesturePageView                                                                                                                              | false                   |
| initialAlignment  | init image rect with alignment when initialScale > 1.0                                                                                                               | InitialAlignment.center |

```dart
ExtendedImage.network(
  imageTestUrl,
  fit: BoxFit.contain,
  //enableLoadState: false,
  mode: ExtendedImageMode.gesture,
  initGestureConfigHandler: (state) {
    return GestureConfig(
        minScale: 0.9,
        animationMinScale: 0.7,
        maxScale: 3.0,
        animationMaxScale: 3.5,
        speed: 1.0,
        inertialSpeed: 100.0,
        initialScale: 1.0,
        inPageView: false,
        initialAlignment: InitialAlignment.center,
        );
  },
)
```

### double tap animation

```dart
onDoubleTap: (ExtendedImageGestureState state) {
  ///you can use define pointerDownPosition as you can,
  ///default value is double tap pointer down postion.
  var pointerDownPosition = state.pointerDownPosition;
  double begin = state.gestureDetails.totalScale;
  double end;

  //remove old
  _animation?.removeListener(animationListener);

  //stop pre
  _animationController.stop();

  //reset to use
  _animationController.reset();

  if (begin == doubleTapScales[0]) {
    end = doubleTapScales[1];
  } else {
    end = doubleTapScales[0];
  }

  animationListener = () {
    //print(_animation.value);
    state.handleDoubleTap(
        scale: _animation.value,
        doubleTapPosition: pointerDownPosition);
  };
  _animation = _animationController
      .drive(Tween<double>(begin: begin, end: end));

  _animation.addListener(animationListener);

  _animationController.forward();
},
```

## Editor

![img](https://github.com/fluttercandies/Flutter_Candies/blob/master/gif/extended_image/editor.gif)

```dart
    ExtendedImage.network(
      imageTestUrl,
      fit: BoxFit.contain,
      mode: ExtendedImageMode.editor,
      extendedImageEditorKey: editorKey,
      initEditorConfigHandler: (state) {
        return EditorConfig(
            maxScale: 8.0,
            cropRectPadding: EdgeInsets.all(20.0),
            hitTestSize: 20.0,
            cropAspectRatio: _aspectRatio.aspectRatio);
      },
    );
```

ExtendedImage

| parameter               | description                                                  | default |
| ----------------------- | ------------------------------------------------------------ | ------- |
| mode                    | image mode (none,gestrue,editor)                             | none    |
| initEditorConfigHandler | init EditorConfig when image is ready.                       | -       |
| extendedImageEditorKey  | key of ExtendedImageEditorState to flip/rotate/get crop rect | -       |

EditorConfig

| parameter              | description                                                        | default                                                      |
| ---------------------- | ------------------------------------------------------------------ | ------------------------------------------------------------ |
| maxScale               | max scale of zoom                                                  | 5.0                                                          |
| cropRectPadding        | the padding between crop rect and image layout rect.               | EdgeInsets.all(20.0)                                         |
| cornerSize             | size of corner shape                                               | Size(30.0, 5.0)                                              |
| cornerColor            | color of corner shape                                              | primaryColor                                                 |
| lineColor              | color of crop line                                                 | scaffoldBackgroundColor.withOpacity(0.7)                     |
| lineHeight             | height of crop line                                                | 0.6                                                          |
| editorMaskColorHandler | call back of editor mask color base on pointerDown                 | scaffoldBackgroundColor.withOpacity(pointerDown ? 0.4 : 0.8) |
| hitTestSize            | hit test region of corner and line                                 | 20.0                                                         |
| animationDuration      | auto center animation duration                                     | Duration(milliseconds: 200)                                  |
| tickerDuration         | duration to begin auto center animation after crop rect is changed | Duration(milliseconds: 400)                                  |
| cropAspectRatio        | aspect ratio of crop rect                                          | null(custom)                                                 |
| initCropRectType       | init crop rect base on initial image rect or image layout rect     | imageRect                                                    |

### crop aspect ratio

it's a double value, so it's easy for you to define by yourself.
following are official values

```dart
class CropAspectRatios {
  /// no aspect ratio for crop
  static const double custom = null;

  /// the same as aspect ratio of image
  /// [cropAspectRatio] is not more than 0.0, it's original
  static const double original = 0.0;

  /// ratio of width and height is 1 : 1
  static const double ratio1_1 = 1.0;

  /// ratio of width and height is 3 : 4
  static const double ratio3_4 = 3.0 / 4.0;

  /// ratio of width and height is 4 : 3
  static const double ratio4_3 = 4.0 / 3.0;

  /// ratio of width and height is 9 : 16
  static const double ratio9_16 = 9.0 / 16.0;

  /// ratio of width and height is 16 : 9
  static const double ratio16_9 = 16.0 / 9.0;
}
```

### crop,flip,reset

- add key for ExtendedImageEditorState

  `final GlobalKey<ExtendedImageEditorState> editorKey =GlobalKey<ExtendedImageEditorState>();`

- rotate right

  `editorKey.currentState.rotate(right: true);`

- rotate left

  `editorKey.currentState.rotate(right: false);`

- flip

  `editorKey.currentState.flip();`

- reset

  `editorKey.currentState.reset();`

### crop data

#### dart library(stable)

- add [Image](https://github.com/brendan-duncan/image) library into your pubspec.yaml, it's used to crop/rotate/flip image data

```yaml
dependencies:
  image: any
```

- get crop rect and raw image data from ExtendedImageEditorState

```dart
  ///crop rect base on raw image
  final Rect cropRect = state.getCropRect();

  var data = state.rawImageData;
```

- convert raw image data to image library data.

```dart
  /// it costs much time and blocks ui.
  //Image src = decodeImage(data);

  /// it will not block ui with using isolate.
  //Image src = await compute(decodeImage, data);
  //Image src = await isolateDecodeImage(data);
  final lb = await loadBalancer;
  Image src = await lb.run<Image, List<int>>(decodeImage, data);
```

- crop,flip,rotate data

```dart
  //clear orientation
  src = bakeOrientation(src);

  if (editAction.needCrop)
    src = copyCrop(src, cropRect.left.toInt(), cropRect.top.toInt(),
        cropRect.width.toInt(), cropRect.height.toInt());

  if (editAction.needFlip) {
    Flip mode;
    if (editAction.flipY && editAction.flipX) {
      mode = Flip.both;
    } else if (editAction.flipY) {
      mode = Flip.horizontal;
    } else if (editAction.flipX) {
      mode = Flip.vertical;
    }
    src = flip(src, mode);
  }

  if (editAction.hasRotateAngle) src = copyRotate(src, editAction.rotateAngle);
```

- convert to original image data

output is raw image data, you can use it to save or any other thing.

```dart
  /// you can encode your image
  ///
  /// it costs much time and blocks ui.
  //var fileData = encodeJpg(src);

  /// it will not block ui with using isolate.
  //var fileData = await compute(encodeJpg, src);
  //var fileData = await isolateEncodeImage(src);
  var fileData = await lb.run<List<int>, Image>(encodeJpg, src);
```

#### native library(faster)

- add [ImageEditor](https://github.com/fluttercandies/flutter_image_editor) library into your pubspec.yaml, it's used to crop/rotate/flip image data

```yaml
dependencies:
  image_editor: any
```

- get crop rect and raw image data from ExtendedImageEditorState

```dart
  ///crop rect base on raw image
  final Rect cropRect = state.getCropRect();

  var data = state.rawImageData;
```

- prepare crop option

```dart
  final rotateAngle = action.rotateAngle.toInt();
  final flipHorizontal = action.flipY;
  final flipVertical = action.flipX;
  final img = state.rawImageData;

  ImageEditorOption option = ImageEditorOption();

  if (action.needCrop) option.addOption(ClipOption.fromRect(rect));

  if (action.needFlip)
    option.addOption(
        FlipOption(horizontal: flipHorizontal, vertical: flipVertical));

  if (action.hasRotateAngle) option.addOption(RotateOption(rotateAngle));
```

- crop with editImage

output is raw image data, you can use it to save or any other thing.

```dart
  final result = await ImageEditor.editImage(
    image: img,
    imageEditorOption: option,
  );
```

[more detail](https://github.com/fluttercandies/extended_image/blob/master/example/lib/common/crop_editor_helper.dart)

## Photo View

ExtendedImageGesturePageView is the same as PageView and it's made for show zoom/pan image.

if you have cache the gesture, remember call clearGestureDetailsCache() method at the right time.(for example,page view page is disposed)

![img](https://github.com/fluttercandies/Flutter_Candies/blob/master/gif/extended_image/photo_view.gif)

ExtendedImageGesturePageView

| parameter    | description              | default |
| ------------ | ------------------------ | ------- |
| cacheGesture | whether should move page | true    |

GestureConfig

| parameter    | description                                                                                                                                                      | default |
| ------------ | ---------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------- |
| cacheGesture | save Gesture state (for example in ExtendedImageGesturePageView, gesture state will not change when scroll back),remember clearGestureDetailsCache at right time | false   |
| inPageView   | whether in ExtendedImageGesturePageView                                                                                                                          | false   |

```dart
ExtendedImageGesturePageView.builder(
  itemBuilder: (BuildContext context, int index) {
    var item = widget.pics[index].picUrl;
    Widget image = ExtendedImage.network(
      item,
      fit: BoxFit.contain,
      mode: ExtendedImageMode.gesture,
      gestureConfig: GestureConfig(
        inPageView: true, initialScale: 1.0,
        //you can cache gesture state even though page view page change.
        //remember call clearGestureDetailsCache() method at the right time.(for example,this page dispose)
        cacheGesture: false
      ),
    );
    image = Container(
      child: image,
      padding: EdgeInsets.all(5.0),
    );
    if (index == currentIndex) {
      return Hero(
        tag: item + index.toString(),
        child: image,
      );
    } else {
      return image;
    }
  },
  itemCount: widget.pics.length,
  onPageChanged: (int index) {
    currentIndex = index;
    rebuild.add(index);
  },
  controller: PageController(
    initialPage: currentIndex,
  ),
  scrollDirection: Axis.horizontal,
),
```

## Slide Out Page

Extended Image support to slide out page as WeChat.

![img](https://raw.githubusercontent.com/fluttercandies/Flutter_Candies/master/gif/extended_image/slide.gif)

### enable slide out page

ExtendedImage


CHANGELOG.md
## [0.9.0]

* Features:
  Add cacheHeight and cacheWidth params for ExtendedImage.network. 
  Add Key extendedImageGestureKey for ExtendedImageGesture.

## [0.8.0]

* Features:
  Add call back [CanScrollPage] for ExtendedImageGesturePageView. 

## [0.7.4]

* Issues:
  Fix ScrollPhysics is not working for ExtendedImageGesturePageView
  
## [0.7.3+1]

* remove docs.

## [0.7.3]

* Improve:
  fix build error on high flutter sdk(> 1.6.0)
  fix analysiz_options

## [0.7.2]

* Features:
  support loading progress for network
  public HttpClient of ExtendedNetworkImageProvider
  public ExtendedImageGestureState for SlideOffsetHandler/SlideEndHandler/SlideScaleHandler to get scale of image
  
## [0.7.1]

* Improve:
  scale parameter of method(handleDoubleTap) is support animationMinScale and animationMaxScale now.  

## [0.7.0]

* Features:
  Support web.
  Add clearMemoryCacheWhenDispose parameter that whether clear memory cache when image is disposed.
* Issues:
  Fix animationMinScale and animationMaxScale are not working for gif.  
  Fix scale parameter of method(handleDoubleTap) is beyond minScale and maxScale.

## [0.6.9]

* Features:
  Support customize offset when slide page.

## [0.6.8]

* Issues:
  Fix breaking change for flutter 1.10.15 about miss load parameter.

## [0.6.7]

* Issues:
  Fix issue that ExtendedImageGesturePageView didn't work well when set initial alignment.

## [0.6.6]

* Features:
  Support init image with alignment when initialScale >1.0.
* Issues:
  Fix issue that scrollDirection didn't work when set it dynamically(ExtendedImageGesturePageView ).
* Improve:
  Add WaterfallFlow demo.

## [0.6.5]

* Features:
  Add completedWidget for ExtendedImageState, it is include gesture or editor, so that you would't miss them
* Improve:
  Improve documents about Load State 

## [0.6.4]

* Issues:
  Fix issue that rawImageData can't be cached for ExtendedExactAssetImageProvider/ExtendedAssetImageProvider.
* Improve:
  Add demo about ImageEditor with native library, it's faster. 

## [0.6.3]

* Issues:
  Fix issue that forget canvas.restore after canvas.clipRect
* Breaking Change:
  ImageEditor：you should crop image data before flip or rotate image data now.
* Improve:
  Increase cropping speed

## [0.6.2]

* Features:
  Add InitCropRectType(imageRect,layoutRect) for EditorConfig to define init crop rect base on initial image rect or image layout rect.
* Breaking Change:
  Make sure the image is all painted to crop,the fit of image must be BoxFit.contain.
    
## [0.6.1]

* Issues:
  Fix issue about drag slowly in ImageEditor

## [0.6.0]

* Issues:
  Fix issue about strange behaviour at slide out page
  
## [0.5.9]

* Add HeroBuilderForSlidingPage call back to fix strange hero animation
  
## [0.5.8]

* Features:
  Support to crop,rotate,flip image
  
## [0.5.6]

* Add key for ExtendedImageSlidePage

## [0.5.5]

* Features:
  Add call back [CanMovePage] for ExtendedImageGesturePageView. [related issue](https://github.com/fluttercandies/extended_image/issues/32) 
  
## [0.5.4]

* Issues:
  Fix issue about borderRadius and border
  Fix demo error about extended_text 

## [0.5.3]

* Improve codes base on v1.7.8

## [0.5.1]

* Features:
  Add call back [onSlidingPage] when is sliding page, you can change other widgets state in page.[ExtendedImageSlidePage]
* Add [enableSlideOutPage] parameter to define whether enable slide out page. [ExtendedImage]  

## [0.4.3]

* Breaking Change:
  Parameter [gestureConfig] is obsolete. [initGestureConfigHandler] is used to setting GestureConfig now.

* Issues:
  Fix issue about slide page.
  
* Features: 
  Support to slide page at loading/failed state

## [0.4.2]

* add README-ZH.md
  
## [0.4.1]

* add SlideType to support slide only image or whole page[ExtendedImageSlidePage]

## [0.4.0]

* support to slide out page

## [0.3.8]

* update path_provider 1.1.0

## [0.3.6]

* handle load failed when re-addListener

## [0.3.4]

* add physics parameter for ExtendedImageGesturePageView

## [0.3.3]

* disabled informationCollector to keep backwards compatibility for now (ExtendedNetworkImageProvider)

## [0.3.2]

* import extended_image_library for network cache

## [0.3.1]

* fix issue that AnimationController.stop() called after AnimationController.dispose().
* show how to build a double tap scale animation.

## [0.2.9]

* add handleDoubleTap method to support zoom image base on double tap position.

## [0.2.8]

* add inertia scroll when image is zoom in and it's moving page.

## [0.2.7]

* fix issue that wrong behavior of page view scroll when image has big width or big height.

## [0.2.6]

* fix issue that wrong behavior of page view scroll when image is zoom in.

## [0.2.5]

* add onDoubleTap parameter to custom double tap behavior under ExtendedImageMode.Gesture

## [0.2.3]

* add enableMemoryCache parameter, whether cache in PaintingBinding.instance.imageCache
* add clearMemoryCacheIfFailed parameter, when failed to load image, whether clear memory cache,if true, image will reload in next time.
* auto cancel network request is obsolete.

## [0.2.2]

* update path_provider version from 0.4.1 to 0.5.0+1

## [0.2.1]

* add cancelToken,retries,timeLimit and timeRetry parameters for ExtendedImage.network method
* add default cancelToken for ExtendedImage.network method
* fix issue about cancel network image request
* fix gesture page view scrolls not smooth

## [0.2.0]

* support zoom/pan image and view image in page view like wechat(support zoom in and scroll next or previous image)

## [0.1.8]

* remove image_picker_saver from extended_image.
  obsolete saveNetworkImageToPhoto method(if you want to save photo,you can import image_picker_saver and get data from getNetworkImageData method)

## [0.1.7]


example/lib/pages/main_page.dart
import 'package:extended_image/extended_image.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:ff_annotation_route/ff_annotation_route.dart';
import 'package:flutter_candies_demo_library/flutter_candies_demo_library.dart';
import 'package:oktoast/oktoast.dart';

import '../example_route.dart';
import '../example_routes.dart' as example_routes;

@FFRoute(
  name: 'fluttercandies://mainpage',
  routeName: 'MainPage',
)
class MainPage extends StatelessWidget {
  MainPage() {
    routeNames.addAll(example_routes.routeNames);
    routeNames.remove('fluttercandies://picswiper');
    routeNames.remove('fluttercandies://mainpage');
    routes.addAll(routeNames
        .map<RouteResult>((String name) => getRouteResult(name: name)));
  }
  final List<RouteResult> routes = <RouteResult>[];
  final List<String> routeNames=<String>[];
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: const Text('ExtendedImage'),
        actions: <Widget>[
          ButtonTheme(
            minWidth: 0.0,
            padding: const EdgeInsets.symmetric(horizontal: 10.0),
            child: FlatButton(
              child: Text(
                'Github',
                style: TextStyle(
                  decorationStyle: TextDecorationStyle.solid,
                  decoration: TextDecoration.underline,
                  color: Colors.white,
                ),
              ),
              onPressed: () {
                launch('https://github.com/fluttercandies/extended_image');
              },
            ),
          ),
          ButtonTheme(
            padding: const EdgeInsets.only(right: 10.0),
            minWidth: 0.0,
            child: FlatButton(
              child:
                  Image.network('https://pub.idqqimg.com/wpa/images/group.png'),
              onPressed: () {
                launch('https://jq.qq.com/?_wv=1027&k=5bcc0gy');
              },
            ),
          )
        ],
      ),
      body: ListView.builder(
        itemBuilder: (BuildContext c, int index) {
          final RouteResult page = routes[index];
          return Container(
              margin: const EdgeInsets.all(20.0),
              child: GestureDetector(
                behavior: HitTestBehavior.translucent,
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: <Widget>[
                    Text(
                      (index + 1).toString() + '.' + page.routeName,
                      //style: TextStyle(inherit: false),
                    ),
                    Text(
                      page.description,
                      style: TextStyle(color: Colors.grey),
                    )
                  ],
                ),
                onTap: () {
                  Navigator.pushNamed(context, routeNames[index]);
                },
              ));
        },
        itemCount: routes.length,
      ),
      floatingActionButton: kIsWeb
          ? null
          : FloatingActionButton(
              onPressed: () {
                ///clear memory
                clearMemoryImageCache();

                ///clear local cahced
                clearDiskCachedImages().then((bool done) {
                  showToast(done ? 'clear succeed' : 'clear failed',
                      position: const ToastPosition(align: Alignment.center));
                });
              },
              child: const Text(
                'clear cache',
                textAlign: TextAlign.center,
                style: TextStyle(
                  inherit: false,
                ),
              ),
            ),
    );
  }
}
